### ISSUE TITLE
Setting pretrained = true causes the cnn_learner model to not train

### ISSUE THREAD

**Issue by butchland** [butchland/fastai_xla_extensions#14]
But setting the pretrained flag to false causes the same model to train quickly
This is shown in the following notebooks:

* [Resnet with pretrained = False -- trains ok ](https://github.com/butchland/fastai_xla_extensions/blob/master/explore_nbs/Basic_resnet_explorationTPU.ipynb)

* [Resnet with pretrained = True -- fails to train](https://github.com/butchland/fastai_xla_extensions/blob/master/explore_nbs/Basic_resnet_explorationTPU_failedtraining_pretrained_true.ipynb)

**Comment by tyoc213** [comment]
I have found that freezing seems to not work???

https://github.com/tyoc213/fastai_xla_extensions/blob/fix_prev_lenet/exploring_freeze_TPU.ipynb

The 2 outputs of the method I found on the forum for print what is freezed and what is not, shows the same output of pretrained is eithere true or false.

In the section of the nb *Manual freezing!* I extract the part of the code that freezes it like

```
n = -1
if tpu_learner_manual.opt is None:
    print("create opttttttttttttttttttttttttttttttttttttttttt")
    tpu_learner_manual.create_opt()
    print(f'create_opt {tpu_learner_manual.opt}')
    tpu_learner_manual.opt = MyXLAOptimProxy(tpu_learner_manual.opt)
tpu_learner_manual.opt.freeze_to(n)
tpu_learner_manual.opt.clear_state()
```

**Comment by butchland** [comment]
Fixed in 0.0.4 
**Root cause analysis:**  
It wasn't the pretrained flag that was causing the problem. It was the fact that the `create_opt` method was called before the `TrainEvalCallback.before_fit` method was called (the `before_fit` method causes the model to be moved to the `learner.dls.device`) So even if the `dls.device` was moved to the TPU device, the `learner.model` hadn't been moved yet, and when the `create_opt` creates an instance of the optimizer using the `learn.model.parameters()` in the optimizer constructor, these parameters were still in the CPU. The additional optimizer hyperparameters and optimizer state were then based of this model parameters. When the model (and its parameters) were then moved later by the `TrainEvalCallback.before_fit` method, the other optimizer state were not moved along with it to the new device (TPU). This kind of behavior seems to be specific to XLA/TPU behavior as this does not cause a problem with moving the parameter to the GPU.
**Fix**
I patched the `Learner.create_opt` and the `Learner.summary` methods (which require both model and dls to interact) so that they first move the model and dls devices to the TPU before doing their usual thing.
This means that loading a pretrained model now will call move the device and model to the TPU since loading a pretrained model from a file will at some point call the `create_opt` method to recreate the optimizer state after loading it from a file.
